src/main/java/com/rapid7/client/dcerpc/io/PacketOutput.java

```diff
@@ -100,29 +110,48 @@ public void writeString(final String string, final boolean nullTerminate)
 
         if (string != null) {
             writeReferentID();
-            writeInt(maximumChars);
-            writeInt(0);
-            writeInt(currentChars);
-            writeChars(string);
-
-            if (nullTerminate) {
-                writeShort((short) 0);
-            }
-
-            align();
+            writeString(string, maximumChars, currentChars, nullTerminate);
         } else {
             writeNull();
         }
     }
 
     public void writeStringRef(final String string, final boolean nullTerminate)
         throws IOException {
-        if (string != null) {
-            writeReferentID();
-            writeString(string, nullTerminate);
+        writeReferentID();
+        writeString(string, nullTerminate);
+    }
+
+    public void writeString(final String string, final boolean nullTerminate)
+        throws IOException {
+        final int maximumChars;
+        final int currentChars;
+
+        if (string == null) {
+            maximumChars = 1;
+            currentChars = 1;
         } else {
-            writeNull();
+            maximumChars = string.length() + (nullTerminate ? 1 : 0);
+            currentChars = string.length() + (nullTerminate ? 1 : 0);
         }
+
+        writeString(string, maximumChars, currentChars, nullTerminate);
+    }
+
+    private void writeString(final String string, final int maximumChars, final int currentChars, final boolean nullTerminate)
+       throws IOException {
+        writeInt(maximumChars); //max_is (max size)
+        writeInt(0); //min_is (offset)
+        writeInt(currentChars); //size_is (actual size)
+        if (string == null) {
+            writeShort(0);
+        } else {
+            writeChars(string);
+        }
+        if (nullTerminate) {
+            writeShort((short) 0);
+        }
+        align();
```

Comment:
Is alignment really necessary here?  This isn't a pointer.

Reply:
Alignment is needed to ensure the `writeShort` does not cause misalignment with the maxCount (test by writing a string with even characters); null termination will require alignment.